### PR TITLE
add probe bof

### PR DESCRIPTION
**Pull request by rwhitcroft**
Test if a remote port is open, similar to `nc -z`.

Issues:
- I defined `struct addrinfo` directly in `bofdefs.h`, but not sure this is allowed. I did this because including `<ws2tcpip.h>` results in warnings from other bofs that include `<windows.h>` before `bofdefs.h`. 
- I don't have a CS license, so I'm using Sliver for all of this. I can't test the `SA.cna` part but tried to copy the other ones.

I realize this is not a groundbreaking addition to the repo, so no big deal if it's too much of a hassle to add. I had a blast learning BOF stuff either way.

```
[server] sliver (ALTERNATIVE_STAMINA) > probe 127.0.0.1 445

[*] Successfully executed probe (coff-loader)
[*] Got output:
127.0.0.1:445 OPEN

[server] sliver (ALTERNATIVE_STAMINA) > probe 127.0.0.1 12345

[*] Successfully executed probe (coff-loader)
[*] Got output:
127.0.0.1:12345 FAILED

[server] sliver (ALTERNATIVE_STAMINA) > probe dc01.ad.local 445

[*] Successfully executed probe (coff-loader)
[*] Got output:
dc01.ad.local:445 OPEN
```

### CODE REVIEW

**Comment by freefirex** [comment]
I'll be happy to review this and work with you on getting it merged in 😄 

Just to let you know timeline I'm teaching this week, so I'll review and provide any feedback monday or tuesday of next week.

File: src/SA/probe/entry.c [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+            struct fd_set sockets;
+            FD_ZERO(&sockets);
+            FD_SET(sock, &sockets);
+
+            WS2_32$connect(sock, result->ai_addr, result->ai_addrlen);
+            WS2_32$select(1, NULL, &sockets, NULL, &tv);
+
+            if (WS2_32$__WSAFDIsSet(sock, &sockets)) {
+                WS2_32$closesocket(sock);
```
Comment (freefirex) [inline src/SA/probe/entry.c:37]: from (https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-closesocket)[https://docs.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-closesocket]
`An application should always have a matching call to closesocket for each successful call to socket to return any socket resources to the system.`

from that we should move this closesocket call below this if statement, rather then inside it

**Review by freefirex — CHANGES_REQUESTED** [review]
This is very good code. I like how you handled timeout / blocking mode.  I've added the changes I'd like made to accept this in, I've tested the changes as written on my end.  

For not having access to sleep you did that side very well as well :)

Thanks again!

**Comment by rwhitcroft** [comment]
Thank you for the review and suggestions! I think I got them all fixed.

**Review by freefirex — APPROVED** [review]

**Comment by freefirex** [comment]
Thanks for making those changes! merged

**Pull request merged**